File: src/main/java/org/reflections/adapters/JavaReflectionAdapter.java
package org.reflections.adapters;

import com.google.common.base.Joiner;
import com.google.common.collect.Lists;
import org.reflections.util.Utils;
import org.reflections.vfs.Vfs.File;

import javax.annotation.Nullable;
import java.lang.annotation.Annotation;
import java.lang.reflect.*;
import java.util.*;

import static org.reflections.ReflectionUtils.forName;

/** */
public class JavaReflectionAdapter implements MetadataAdapter<Class, Field, Member> {

    @Override
    public List<Field> getFields(Class cls) {
        return Lists.newArrayList(cls.getDeclaredFields());
    }

    @Override
    public List<Member> getMethods(Class cls) {
        List<Member> methods = Lists.newArrayList();
        methods.addAll(Arrays.asList(cls.getDeclaredMethods()));
        methods.addAll(Arrays.asList(cls.getDeclaredConstructors()));
        return methods;
    }

    @Override
    public String getMethodName(Member method) {
        return (method instanceof Method) ? method.getName() : ((method instanceof Constructor) ? "<init>" : null);
    }

    @Override
    public List<String> getParameterNames(Member member) {
        List<String> result = Lists.newArrayList();

        Class<?>[] parameterTypes = (member instanceof Executable) ? ((Executable) member).getParameterTypes() : null;

        if (parameterTypes != null) {
            for (Class<?> paramType : parameterTypes) {
                String name = getName(paramType);
                result.add(name);
            }
        }

        return result;
    }

    @Override
    public List<String> getClassAnnotationNames(Class aClass) {
        return getAnnotationNames(aClass.getDeclaredAnnotations());
    }

    @Override
    public List<String> getFieldAnnotationNames(Field field) {
        return getAnnotationNames(field.getDeclaredAnnotations());
    }

    @Override
    public List<String> getMethodAnnotationNames(Member method) {
        Annotation[] annotations = (method instanceof Executable)
                                   ? ((Executable) method).getDeclaredAnnotations()
                                   : null;
        return getAnnotationNames(annotations);
    }

    @Override
    public List<String> getParameterAnnotationNames(Member method, int parameterIndex) {
        Annotation[][] annotations = (method instanceof Executable)
                                     ? ((Executable) method).getParameterAnnotations()
                                     : null;

        return (annotations != null) ? getAnnotationNames(annotations[parameterIndex]) : getAnnotationNames(null);
    }

    @Override
    public String getReturnTypeName(Member method) {
        return ((Method) method).getReturnType().getName();
    }

    @Override
    public String getFieldName(Field field) {
        return field.getName();
    }

    @Override
    public Class getOrCreateClassObject(File file) {
        return getOrCreateClassObject(file, null);
    }

    public static Class getOrCreateClassObject(File file, @Nullable ClassLoader... loaders) {
        String name = file.getRelativePath().replace("/", ".").replace(".class", "");
        return forName(name, loaders);
    }

    @Override
    public String getMethodModifier(Member method) {
        return Modifier.toString(method.getModifiers());
    }

    @Override
    public String getMethodKey(Class cls, Member method) {
        return getMethodName(method) + '(' + Joiner.on(", ").join(getParameterNames(method)) + ')';
    }

    @Override
    public String getMethodFullKey(Class cls, Member method) {
        return getClassName(cls) + '.' + getMethodKey(cls, method);
    }

    @Override
    public boolean isPublic(Object o) {
        Integer mod = (o instanceof Class)
                      ? ((Class) o).getModifiers()
                      : ((o instanceof Member) ? ((Member) o).getModifiers() : null);

        return (mod != null) && Modifier.isPublic(mod);
    }

    @Override
    public String getClassName(Class cls) {
        return cls.getName();
    }

    @Override
    public String getSuperclassName(Class cls) {
        Class superclass = cls.getSuperclass();
        return (superclass != null) ? superclass.getName() : "";
    }

    @Override
    public List<String> getInterfacesNames(Class cls) {
        Class[]      classes = cls.getInterfaces();
        List<String> names;
        names = (classes != null) ? new ArrayList<>(classes.length) : new ArrayList<>(0);
        if (classes != null) {
            for (Class cls1 : classes) {
                names.add(cls1.getName());
            }
        }
        return names;
    }

    @Override
    public boolean acceptsInput(String file) {
        return file.endsWith(".class");
    }

    //
    private static List<String> getAnnotationNames(Annotation[] annotations) {
        List<String> names = new ArrayList<>(annotations.length);
        for (Annotation annotation : annotations) {
            names.add(annotation.annotationType().getName());
        }
        return names;
    }

    public static String getName(Class type) {
        if (type.isArray()) {
            try {
                Class cl  = type;
                int   dim = 0;
                while (cl.isArray()) {
                    dim++;
                    cl = cl.getComponentType();
                }
                return cl.getName() + Utils.repeat("[]", dim);
            } catch (Throwable e) {
                //
            }
        }
        return type.getName();
    }
}

File: src/main/java/org/reflections/adapters/JavaReflectionAdapter.kt
package org.reflections.adapters

import org.reflections.util.classForName
import org.reflections.util.generateWhileNotNull
import org.reflections.util.tryOrDefault
import org.reflections.vfs.VfsFile
import java.lang.reflect.Constructor
import java.lang.reflect.Executable
import java.lang.reflect.Field
import java.lang.reflect.Member
import java.lang.reflect.Method
import java.lang.reflect.Modifier

val CreateJavaReflectionClassAdapter = { vfsFile: VfsFile ->
    JavaReflectionClassAdapter(classForName(vfsFile.relativePath!!.replace("/", ".").replace(".class", ""))!!)
}

data class JavaReflectionClassAdapter(val cls: Class<*>) : ClassAdapter {
    override val fields = cls.declaredFields.map { JavaReflectionFieldAdapter(it) }
    override val methods = (cls.declaredMethods.asList() + cls.declaredConstructors).map {
        JavaReflectionMethodAdapter(it)
    }

    override val annotations = cls.declaredAnnotations.map { it.annotationClass.java.name }
    override val name = cls.name!!
    override val superclass = cls.superclass?.name.orEmpty()
    override val interfaces = cls.interfaces.map { it.name!! }
    override val isPublic = Modifier.isPublic(cls.modifiers)
}

data class JavaReflectionFieldAdapter(val field: Field) : FieldAdapter {
    override val annotations = field.declaredAnnotations.map { it.annotationClass.java.name }
    override val name = field.name!!
    override val isPublic = Modifier.isPublic(field.modifiers)
}

data class JavaReflectionMethodAdapter(val method: Member) : MethodAdapter {
    override val name = when (method) {
        is Method         -> method.name
        is Constructor<*> -> "<init>"
        else              -> TODO()
    }

    override val parameters = (method as? Executable)?.parameterTypes.orEmpty().map { cls ->
        when {
            cls.isArray -> tryOrDefault(cls.name) {
                val componentTypes = cls.generateWhileNotNull { componentType }
                componentTypes.last().name + "[]".repeat(componentTypes.size - 1)
            }
            else        -> cls.name
        }!!
    }

    override val annotations = (method as Executable).declaredAnnotations!!.map { it.annotationClass.java.name }

    override fun parameterAnnotations(parameterIndex: Int) =
            (method as Executable).parameterAnnotations[parameterIndex].map { it.annotationClass.java.name }

    override val returnType = (method as Method).returnType.name!!
    override val modifier = Modifier.toString(method.modifiers)!!
    override val isPublic = Modifier.isPublic(method.modifiers)
}